src/libexpr/flake/call-flake.nix

```diff
@@ -12,7 +12,36 @@ overrides:
 
 let
 
-  lockFile = builtins.fromJSON lockFileStr;
+  origLockFile = builtins.fromJSON lockFileStr;
+
+  # Build an optimized lock file by consolidating equivalent inputs.
+  lockFile =
+    let
+      nodeCacheKey = key: node:
+        let taint = if overrides ? ${key} then { taintedAs = key; } else {};
+        in builtins.toJSON (node.locked or {} // taint);
```

Comment:
There's probably a better way to do this than `builtins.toJSON`, but this works as a proof of concept.

Reply:
Thinking about this some more, using `toJSON` seems hacky at first, but it's actually probably more performant than trying to construct a custom string key with nix.